init.lua

```diff
@@ -1333,20 +1338,162 @@ function torchcraft:get_layer(layer_type, also_enemy)
                 end
-                img = self:draw_entity(img, pos_x, pos_y,
-                                       ut.pixel_size_x, ut.pixel_size_y,
-                                       color)
+                if color then
+                    img = self:draw_entity(img, pos_x, pos_y,
+                                           ut.pixel_size_x, ut.pixel_size_y,
+                                           color)
+                end
+            end
+        end
+    end
+    return img
+end
+
+function torchcraft:draw_value(img, pos_x, pos_y, size_x, size_y, value)
+    local narrow_width_start = math.max(1,
+                                        math.floor(pos_x - size_x / 2))
+    local narrow_width_end = math.min(self.field_size[1],
+                                      math.floor(pos_x + size_x / 2))
+    local narrow_height_start = math.max(1,
+                                         math.floor(pos_y - size_y / 2))
+    local narrow_height_end = math.min(self.field_size[2],
+                                       math.floor(pos_y + size_y / 2))
+    -- Hack to fix crash were unit is on top left corner of FoV
+    if narrow_width_end <= 0 then
+        narrow_width_end = narrow_width_start
+    end
+    local rect = img:narrow(2, narrow_width_start,
+                            narrow_width_end - narrow_width_start + 1)
+    if narrow_height_end <= 0 then
+        narrow_height_end = narrow_height_start
+    end
+    local rect = rect:narrow(1, narrow_height_start,
+                             narrow_height_end - narrow_height_start + 1)
+    rect:fill(value)
+    return img
+end
+
+function torchcraft:get_feature(feature, also_enemy, dtype)
+--returns 2D imagelike tensor of feature
+--type defaults to ByteTensor
+
+-- :TODO: Maybe make categorical features return a 3D tensor of one-hots
+--        Or just render them to colors with draw value to increase spacing
+
+    local img
+    if dtype == "short" then
+        img = torch.ShortTensor(self.field_size[2],self.field_size[1]):fill(0)
+    elseif dtype == "int" then
+        img = torch.IntTensor(self.field_size[2],self.field_size[1]):fill(0)
+    elseif dtype == "long" then
+        img = torch.LongTensor(self.field_size[2],self.field_size[1]):fill(0)
+    elseif dtype == "float" then
+        img = torch.FloatTensor(self.field_size[2],self.field_size[1]):fill(0)
+    elseif dtype == "double" then
+        img = torch.DoubleTensor(self.field_size[2],self.field_size[1]):fill(0)
+    else
+        img = torch.ByteTensor(self.field_size[2],self.field_size[1]):fill(0)
+    end
+
+    local pos_x, pos_y
+
+    if feature == "visibility" then
+        for y, xs in ipairs(self.state.visibility) do
+            for x, value in ipairs(xs) do
+                pos_x = (x - 1) * 32
+                pos_y = (y - 1) * 32
+                img = self:draw_value(img, pos_x, pos_y,
+                                       32, 32, value)
+            end
+        end
+        return img
```

Comment:
I cleaned this up some. I'll probably change it some more once I see how you're handling the img variable.